src/cockpit/polkit.py

```diff
@@ -0,0 +1,139 @@
+                    prompt = ''
+
+                try:
+                    passwd = await self.responder.do_askpass('', prompt, '')
+                    logger.debug('got PAM_PROMPT %s, do_askpass returned a password', prompt)
+                except asyncio.CancelledError:
+                    logger.debug('got PAM_PROMPT %s, do_askpass cancelled', prompt)
+                    process.terminate()
+                    break
+                process.stdin.write(passwd.encode())
+                process.stdin.write(b'\n')
+                del passwd  # don't keep this around longer than necessary
+                await process.stdin.drain()
+                logger.debug('got PAM_PROMPT, wrote password to helper')
+            elif line.startswith(b'PAM_TEXT_INFO'):
+                logger.info(line.decode())
+            elif line.startswith(b'PAM_ERROR'):
+                logger.warning(line.decode())
+            elif line.startswith(b'SUCCESS'):
+                logger.info('Authentication succeeded')
+                break
+            elif line.startswith(b'FAILURE'):
+                logger.warning('Authentication failed')
+                break
+            else:
+                logger.warning('Unknown line from helper, aborting: %s', line)
+                process.terminate()
+                break
+
+        res = await process.wait()
+        logger.debug('helper exited with code %i', res)
+
+    @bus.Interface.Method('', 's')
+    def cancel_authentication(self, cookie):
+        logger.warning('CancelAuthentication: %s not implemented', cookie)
+
+
+def register_agent(responder):
+    '''Register polkit agent.
+
+    Returns an opaque object which needs to be kept during the lifetime of the agent.
+    After that, pass it to @unregister_agent().
+    '''
+    try:
+        system_bus = bus.Bus.default_system()
+    except FileNotFoundError:
+        logger.debug('system bus not found, not registering polkit agent')
+        return None
+
+    try:
+        # may refine that with a D-Bus call to logind
+        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+    except KeyError:
+        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+        return None
+
+    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
+    agent_slot = system_bus.add_object('/', agent_object)
+
+    # register agent
+    (loc, encoding) = locale.getlocale()
+    locale_name = loc or 'C'
+    if encoding:
+        locale_name += '.' + encoding
+    system_bus.call_method(
+        'org.freedesktop.PolicyKit1',
+        '/org/freedesktop/PolicyKit1/Authority',
+        'org.freedesktop.PolicyKit1.Authority',
+        'RegisterAuthenticationAgent',
+        '(sa{sv})ss',
+        subject, locale_name, '/')
+    logger.debug('Registered agent for %r and locale %s', subject, locale_name)
+    return {'subject': subject, 'slot': agent_slot}
+
+
+def unregister_agent(agent):
```

Comment:
Do this as an asynchronous context manager.

Reply:
Note: [contextlib.asyncccontextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager) is Python ≥ 3.7, so this needs to be done without the helper as class. Also, this is a bit awkward: I actually only want to register the agent for pkexec bridges, so `do_connect_transport()` would need two very similar code paths, one for the polkit context manager, one for other superuser peers.

So I dropped the "meh" comment with the `if pkexec in args` check, and just always register an agent for any superuser bridge. That's unnecessary for sudo, but perhaps it helps other "become root" implementations.

I  pushed this as a separate fixup commit for the time being, so that it's easier to review/discuss/possibly revert.

Reply:
I wonder if you could try to do it non-asynchronously, then, using a normal context manager.  The startup needs to be done async, but the teardown could be done 'fire and forget' — you don't at all care about the reply to the "unregister" message...

Something like:

```python
@contextlib.contextmanager
def yield_agent(agent):
    try:
        yield agent
    finally:
        ... shutdown here ...

async def register_agent(responder):
    await ... do stuff ...
    return yield_agent(agent)

... and elsewhere ...
async def do_connect_transport(self):
    ...
    with await polkit.register_agent():
        do stuff...
    ....
```

*shrug*  It's slightly more elegant than what you have here, and more robust ,but also more confusing and perhaps not really worth the effort.

Reply:
I initially liked that, then started to implement that, but only just now fully realized what you meant. Doing an async D-Bus call with "fire and forget" is prone to run into uncaught exceptions, TBH I'm not a big fan of that.

I reworked this to not run an agent for sudo (i.e. "not pkexec"), though.